specs/rollup-node.md

```diff
@@ -175,58 +235,77 @@ The following scenarios are assimilated to errors:
+# Handling L1 Re-Orgs
 
-[l1-finality]: https://www.paradigm.xyz/2021/07/ethereum-reorgs-after-the-merge/
+[l1-reorgs]: #handling-L1-re-orgs
+
+The [previous section on L2 chain derivation][l2-chain-derivation] assumes linear progression of the L1 chain. It is
+also applicable for batch processing, meaning that any given point in time, the canonical L2 chain is given by
+processing the whole L1 chain since the [L2 chain inception][g-inception].
+
+> By itself, the previous section fully specifies the behaviour of the rollup driver. **The current section is
+> non-specificative** but shows how L1 re-orgs can be handled in practice.
+
+In practice, the L1 chain is processed incrementally. However, the L1 chain may occasionally [re-organize][g-reorg],
+meaning the head of the L1 chain changes to a block that is not the child of the previous head but rather another
+descendant of an ancestor of the previous head. In that case, the rollup driver must first search for the common L1
+ancestor, and can re-derive the L2 chain from that L1 block and onwards.
+
+The starting point of the re-derivation is a pair `(refL2, nextRefL1)` where `refL2` refers to the L2 block to build
+upon and `nextRefL1` refers to the next L1 block to derive from (i.e. if `refL2` is derived from L1 block `refL1`,
+`nextRefL1` is the canonicla L1 block at height `l1Number(refL1) + 1`).
+
+In practice, the happy path (no re-org) and the re-org paths are merged. The happy path is simply a special case of the
+re-org path where the starting point of the re-derivation is `(currentL2Head, newL1Block)`.
+
+This re-derivation starting point can be found by applying the following algorithm:
+
+1. (Initialization) Set the initial `refL2` to the head block of the L2 execution engine.
+2. Set `parentL2` to `refL2`'s parent block and `refL1` to the L1 block that `refL2` was derived from.
+3. Fetch `currentL1`, the canonical L1 block at the same height as `refL1`.
+
+- If `currentL1 == refL1`, then `refL2` was built on a canonical L1 block:
+  - Find the next L1 block (it may not exist yet) and return `(refL2, nextRefL1)` as the starting point of the
+    re-derivation.
+    - It is necessary to ensure that no L1 re-org occured during this lookup, i.e. that `nextRefL1.parent == refL1`.
+    - If the next L1 block does not exist yet, there is no re-org, and nothing new to derive, and we can abort the
+        process.
+- Otherwise, if `refL2` is the L2 genesis block, we have re-orged past the genesis block, which is an error that
+    requires a re-genesis of the L2 chain to fix (i.e. creating a new genesis configuration).
+- Otherwise, if either `currentL1` does not exist, or `currentL1 != refL1`, set `refL2` to `parentL2` and restart this
+    algorithm from step 2.
+  - Note: if `currentL1` does not exist, it means we are in a re-org to a shorter L1 chain.
+  - Note: as an optimization, we can cache `currentL1` and reuse it as the next value of `nextRefL1` to avoid an
+        extra lookup.
```

Comment:
I've tried simplifying this as much as possible while preserve the semantics. @protolambda could you look it over to see if it seems alright?

Reply:
New version looks good!